src/ConnectedMode/BranchMatcher.cs

```diff
@@ -51,64 +52,85 @@ public interface IBranchMatcher
-                {
-                    return head.FriendlyName;
-                }
+            if (head == null)
+            {
+                logger.LogVerbose(Resources.BranchMapper_NoHead);
+                return null;
+            }
 
-                string closestBranch = null;
-                int closestDistance = int.MaxValue;
+            var remoteBranches = await sonarQubeService.GetProjectBranchesAsync(projectKey, token);
 
-                Lazy<Commit[]> headCommits = new Lazy<Commit[]>(() => head.Commits.ToArray());
+            if (remoteBranches.Any(rb => string.Equals(rb.Name, head.FriendlyName, StringComparison.InvariantCultureIgnoreCase)))
+            {
+                logger.LogVerbose(Resources.BranchMapper_Match_SameSonarBranchName, head.FriendlyName);
+                return head.FriendlyName;
+            }
 
-                foreach (var remoteBranch in remoteBranches)
-                {
-                    var localBranch = gitRepo.Branches.FirstOrDefault(r => string.Equals(r.FriendlyName, remoteBranch.Name, StringComparison.InvariantCultureIgnoreCase));
+            string closestBranch = null;
+            int closestDistance = int.MaxValue;
 
-                    if (localBranch == null) { continue; }
+            Lazy<Commit[]> headCommits = new Lazy<Commit[]>(() => head.Commits.ToArray());
 
-                    var distance = GetDistance(headCommits.Value, localBranch, closestDistance);
+            logger.LogVerbose(Resources.BranchMapper_CheckingSonarBranches);
+            foreach (var remoteBranch in remoteBranches)
+            {
 
-                    if (distance < closestDistance)
-                    {
-                        closestBranch = localBranch.FriendlyName;
-                        closestDistance = distance;
-                    }
-                }
+                var localBranch = gitRepo.Branches.FirstOrDefault(r => string.Equals(r.FriendlyName, remoteBranch.Name, StringComparison.InvariantCultureIgnoreCase));
 
-                if (closestBranch == null)
+                if (localBranch == null) { continue; }
+
+                var distance = GetDistance(headCommits.Value, localBranch, closestDistance);
+
+                if (distance < closestDistance)
                 {
-                    return remoteBranches.First(rb => rb.IsMain).Name;
+                    closestBranch = localBranch.FriendlyName;
+                    closestDistance = distance;
+                    logger.LogVerbose(Resources.BranchMapper_UpdatingClosestMatch, closestBranch, distance);
                 }
-                return closestBranch;
             }
-            finally
+
+            if (closestBranch == null)
             {
-                CodeMarkers.Instance.GetMatchingBranchStop();
+                logger.LogVerbose(Resources.BranchMapper_NoMatchingBranchFound);
+                closestBranch = remoteBranches.First(rb => rb.IsMain).Name;
```

Comment:
Actually, I did change line 130 to set `closestBranch` rather than returning directly.

I did this before I created the new internal `Do` method, which made the change unnecessary. However, both versions work.